Add SQL Enum support

michnovka opened the pull request:
Backport the feature originally present in 1.x for native ENUM

michnovka committed:
Added .idea into gitignore

michnovka committed:
Added required dependencies for running PHPUnit

michnovka committed:
Added support for ENUM classes

michnovka commented:
I have made some changes to .gitignore, to ignore .idea files (I assume I cant be the only one working in PHPStorm) as its standard used by both Symfony and Doctrine.

I have also included phpunit and psalm dependencies as dev, because well we have lot of @psalm annotations, and then because we actually do have tests for this bundle, so it makes no sense not to have phpunit included.

Speaking of tests, I would appreciate some help on writing them, as I am still a bit lost as to how this bundle is structured. We can start by writing down list of tests that need to be implemented and move on from there:

1. Create schema with ENUM type
2. Check retrieval and cast of values from SQL DB ENUM type into PHP representation
3. Check saving of values from PHP into SQL DB
4. Check configuration option `doctrine.type.enum_sql_declaration` works as expected (enabled ENUM SQLs by default)
5. Check configuration option doctrine.type.types.type = enum works as expected

Possible considerations/discussion:
1. I have made the option to use INT ENUMs when its BackedEnum of int type. I.e. to have ENUM(1,2,3) instead of ENUM('1','2','3'). Yet if its FlagBag, I keep the INT type in DB. Is this desired? For that reason I had to change `Elao\Enum\Bridge\Doctrine\DBAL\Types\AbstractEnumType::isIntBackedEnum` from private to protected.
2. Do we want ODM support too? I dont use ODM in my life, so I have literally 0 experience with it.
3. Are we happy with the type name `SINGLE` ? Maybe using `PRIMITIVE` or `SIMPLE` has more sense?
4. Before we make the final release of 2.x maybe it would be interesting to consider removing all `const` values used inside the bundle into their own Enums? After all this is an enum package :) e.g. `Elao\Enum\Bridge\Doctrine\DBAL\Types\TypesDumper::TYPE_SINGLE` could be made into an Enum type. Checked the code for `const`s and `TypeDumper` seems to be the only place, so I could make that change as part of this PR

ogizanagi commented:
> I have made some changes to .gitignore, to ignore .idea files (I assume I cant be the only one working in PHPStorm) as its standard used by both Symfony and Doctrine.

You're not the only one 😉 But the good practice actually is to register these specific files in a global `.gitignore` file in your user's home directory, not the project one. See https://gist.github.com/subfuzion/db7f57fff2fb6998a16c for instance.

> I have also included phpunit and psalm dependencies as dev, because well we have lot of https://github.com/psalm annotations, and then because we actually do have tests for this bundle, so it makes no sense not to have phpunit included.

Please revert these changes, since these are not necessary. We already have PHPUnit installed through the Symfony phpunit-bridge. Use `vendor/bin/simple-phpunit` to execute the tests.
If you have `make` and Symfony CLI installed on your machine, you can run `make test`. See the [Makefile](https://github.com/Elao/PhpEnums/blob/2.x/Makefile)

About Psalm, the annotations were added for using psalm/phpstan on userland code, we do not have any CI task enforcing it yet ; it should be added in another PR along with proper config & CI.

---

About tests, some existing resources:


>1. Create schema with ENUM type

The [integration tests](https://github.com/Elao/PhpEnums/tree/2.x/tests/Integration) are using a Symfony app to configure some enum using the bundle and the ORM config:

https://github.com/Elao/PhpEnums/blob/2.x/tests/Fixtures/Integration/Symfony/config/config.yml#L19-L33

Then, an integration test case class using the Symfony Kernel make some assertions by consuming the app:

https://github.com/Elao/PhpEnums/blob/2.x/tests/Integration/Bridge/Doctrine/DBAL/Type/EnumTypeTest.php

There are also unit tests for this and the following points for:
- [the DBAL type itself](https://github.com/Elao/PhpEnums/blob/2.x/tests/Unit/Bridge/Doctrine/DBAL/Types/EnumTypeTest.php)
- the [TypesDumper generated code](https://github.com/Elao/PhpEnums/blob/2.x/tests/Unit/Bridge/Doctrine/DBAL/Types/TypesDumperTest.php)

>2. Check retrieval and cast of values from SQL DB ENUM type into PHP representation

See [this integration test](https://github.com/Elao/PhpEnums/blob/49a73e0d11684318e11ad968731a83b79f55fead/tests/Integration/Bridge/Doctrine/DBAL/Type/EnumTypeTest.php#L44-L54) and [unit test](https://github.com/Elao/PhpEnums/blob/2.x/tests/Unit/Bridge/Doctrine/DBAL/Types/EnumTypeTest.php)

>3. Check saving of values from PHP into SQL DB

See [this integration test](https://github.com/Elao/PhpEnums/blob/49a73e0d11684318e11ad968731a83b79f55fead/tests/Integration/Bridge/Doctrine/DBAL/Type/EnumTypeTest.php#L56-L69) and [unit test](https://github.com/Elao/PhpEnums/blob/2.x/tests/Unit/Bridge/Doctrine/DBAL/Types/EnumTypeTest.php)

>4. Check configuration option `doctrine.type.enum_sql_declaration` works as expected (enabled ENUM SQLs by default)

Such tests are made with unit tests:

- [testing the Configuration class](https://github.com/Elao/PhpEnums/blob/49a73e0d11684318e11ad968731a83b79f55fead/tests/Unit/Bridge/Symfony/Bundle/DependencyInjection/ConfigurationTest.php#L36-L58)
- [testing the bundle Extension](https://github.com/Elao/PhpEnums/blob/49a73e0d11684318e11ad968731a83b79f55fead/tests/Unit/Bridge/Symfony/Bundle/DependencyInjection/ElaoEnumExtensionTest.php#L35-L64)

>5. Check configuration option doctrine.type.types.type = enum works as expected

this should be tested in both unit tests case mentioned below (config & extension).

---

> 1. I have made the option to use INT ENUMs when its BackedEnum of int type. I.e. to have ENUM(1,2,3) instead of ENUM('1','2','3'). Yet if its FlagBag, I keep the INT type in DB. Is this desired? For that reason I had to change `Elao\Enum\Bridge\Doctrine\DBAL\Types\AbstractEnumType::isIntBackedEnum` from private to protected.

I think sql enum should preferably be stored as string. So let's always use `ENUM('1','2','3')`.

About `FlagBag`, you should not worry about it in this PR, since it would be a dedicated DBAL type storing the combination of multiple enum cases into 1 INT column. So you cannot store it using an `ENUM` SQL type.

> 2. Do we want ODM support too? I dont use ODM in my life, so I have literally 0 experience with it.

ODM is already supported since #176. There is no change to made in tis PR regarding a "native" enum type in MongoDB.

> 3. Are we happy with the type name `SINGLE` ? Maybe using `PRIMITIVE` or `SIMPLE` has more sense?

I'd go with `SCALAR`

4. Before we make the final release of 2.x maybe it would be interesting to consider removing all `const` values used inside the bundle into their own Enums? After all this is an enum package :) e.g. `Elao\Enum\Bridge\Doctrine\DBAL\Types\TypesDumper::TYPE_SINGLE` could be made into an Enum type. Checked the code for `const`s and `TypeDumper` seems to be the only place, so I could make that change as part of this PR

That's a bit out-of-scope, so I'd say please avoid doing this in this PR 😄 Anyway, keep in mind not every const values make sense as enums. Regarding the TypesDumper ones, it doesn't bring much value, mostly because it's internal as well as using enums instead might complexify a bit the Symfony DI config handling which does not support backed enums out-of-the-box.

michnovka committed:
Removed dev dependencies. Do not use ENUMs with int types, only string ones

michnovka commented:
Good to know about that global .gitignore, I learnt something new.

I have removed the newly added dependencies, except for doctrine-odm-bundle. This is vital for phpunit tests of ODM. If I dont add it, I always get errors when running tests. Psalm and phpunit gone.                                  

I have removed int ENUMs, so they are always strings.

Did not yet rename SINGLE to SCALAR, I will do so once we agree on it globally. Or does your ring rule them all @ogizanagi ? :)

Will have a look at the remaining points and mainly tests later.

michnovka committed:
Revert some changes from initial commits in the PR

ogizanagi commented:
>  have removed the newly added dependencies, except for doctrine-odm-bundle. This is vital for phpunit tests of ODM

That's unexpected : the test cases should be skipped when the bundle is not installed 🤔 :

https://github.com/Elao/PhpEnums/blob/49a73e0d11684318e11ad968731a83b79f55fead/tests/Integration/Bridge/Doctrine/ODM/Type/EnumTypeTest.php#L28-L30

> Did not yet rename SINGLE to SCALAR, I will do so once we agree on it globally. 

Let's go with `SCALAR` for DBAL types yes (since we'll also have `enum` and later on: `flagbag`, `json_collection`, `csv_collection`). For ODM ones, we'll keep `SINGLE` one as opposed to `COLLECTION`, as already implemented and since there should not be any other value.

michnovka committed:
Removed doctrine/mongodb-odm-bundle

michnovka commented:
> That's unexpected : the test cases should be skipped when the bundle is not installed thinking

Ah, I didnt realize its test skip not an error. So this is desired? I think that if we have tests for that part of bridge, then why not require it as dev? Anyways removing since I am not touching ODM so it should not be part of this PR

ogizanagi commented:
> > That's unexpected : the test cases should be skipped when the bundle is not installed thinking
> 
> Ah, I didnt realize its test skip not an error. So this is desired? I think that if we have tests for that part of bridge, then why not require it as dev? Anyways removing since I am not touching ODM so it should not be part of this PR

The reasoning here simply is testing ODM requires an advanced setup on your machine (php mongodb extension, a mongodb instance running, …), so we won't impose it be default and simply skip the related test cases when the ODM bundle is not installed (it's installed on CI in the github actions, so any PR is fully tested).

michnovka commented:
Working on the tests and I have some confusion in https://github.com/Elao/PhpEnums/blob/49a73e0d11684318e11ad968731a83b79f55fead/tests/Integration/Bridge/Doctrine/DBAL/Type/EnumTypeTest.php#L56

Why does saving a `Card` entity with `Card.suit = null` result in `Spades` in DB and on retrieval? The `Card.suit` field is nullable, null is therefore an accepted value

michnovka commented:
I came across one "issue" or better said property of current `TypesDumper`.

I tried this config:
```yaml
elao_enum:
    doctrine:
        types:
            App\Enum\Suit:
                default: !php/const App\Enum\Suit::Spades
            suit_sql_enum:
                class: App\Enum\Suit
                type: enum
```

So I wanted to have an alias for the same Enum. This resulted in this code by `TypesDumper`

```php
<?php

namespace ELAO_ENUM_DT_DBAL\App\Enum {

    if (!\class_exists(SuitType::class)) {
        class SuitType extends \Elao\Enum\Bridge\Doctrine\DBAL\Types\AbstractEnumType
        {
            protected function getEnumClass(): string
            {
                return \App\Enum\Suit::class;
            }

            protected function onNullFromDatabase(): ?\BackedEnum
            {
                return \App\Enum\Suit::from('S');
            }

            protected function onNullFromPhp(): int|string|null
            {
                return 'S';
            }
        }
    }

    if (!\class_exists(SuitType::class)) {
        class SuitType extends \Elao\Enum\Bridge\Doctrine\DBAL\Types\AbstractEnumSQLDeclarationType
        {
            protected function getEnumClass(): string
            {
                return \App\Enum\Suit::class;
            }

            public function getName(): string
            {
                return 'suit_sql_enum';
            }
        }
    }

}
```

So the issue is clear. The name for the types is based on the Enum class name. I think it should be based instead on the type name as specified in config and slugged. Opinions?

ogizanagi commented:
You're right indeed, using the PascalCased type name might be better for such edge-cases.
Just duplicate the enum class for your test-case if you need it for now, in case you prefer not handling with this in this PR right now. We'll fix it in a dedicated one later.

michnovka commented:
I am fine with doing it in this one. Can we use https://www.doctrine-project.org/projects/doctrine-orm/en/latest/reference/namingstrategy.html since Doctrine is needed anyways? Or do you prefer I make my own toPascalCase function?

michnovka commented:
@ogizanagi Running into issues with tests with SQLite... It has no ENUM type. What do you suggest? I have checked the original test in 1.x and there seems to be no testing of actual SQLEnum type functionality

michnovka commented:
There are further issues with this. Because the test requires use of SchemaManager to create the DB, I am getting the following errors when making the tests:

```
Doctrine\DBAL\Exception : Unknown database type enum requested, Doctrine\DBAL\Platforms\MariaDb1027Platform may not support it.
 /home/superuser/scripts/PhpEnums/vendor/doctrine/dbal/src/Platforms/AbstractPlatform.php:396
 /home/superuser/scripts/PhpEnums/vendor/doctrine/dbal/src/Schema/MySQLSchemaManager.php:137
 /home/superuser/scripts/PhpEnums/vendor/doctrine/dbal/src/Schema/AbstractSchemaManager.php:981
 /home/superuser/scripts/PhpEnums/vendor/doctrine/dbal/src/Schema/AbstractSchemaManager.php:208
 /home/superuser/scripts/PhpEnums/vendor/doctrine/dbal/src/Schema/AbstractSchemaManager.php:321
 /home/superuser/scripts/PhpEnums/vendor/doctrine/dbal/src/Schema/MySQLSchemaManager.php:336
 /home/superuser/scripts/PhpEnums/vendor/doctrine/dbal/src/Schema/AbstractSchemaManager.php:306
 /home/superuser/scripts/PhpEnums/vendor/doctrine/dbal/src/Schema/AbstractSchemaManager.php:1229
 /home/superuser/scripts/PhpEnums/vendor/doctrine/orm/lib/Doctrine/ORM/Tools/SchemaTool.php:846
 /home/superuser/scripts/PhpEnums/vendor/doctrine/orm/lib/Doctrine/ORM/Tools/SchemaTool.php:831
 /home/superuser/scripts/PhpEnums/tests/Integration/Bridge/Doctrine/DBAL/Type/EnumTypeTest.php:33
 ```

I think this is the reason why we didnt have tests in 1.x. I am not sure what we can do fix this, any ideas?

michnovka committed:
Renamed SINGLE to SCALAR. Changed TypesDumper class name deduction. Added tests

michnovka committed:
Add comments on TypesDumper and pascalcase only when custom name is used

michnovka commented:
I have pushed the changes so you can have a look too. But I dont think I can resolve this. The original version also did not have tests and I believe the main issue is due to SQLite lack of support. However even if we use locally MySQL, I have the problems with SchemaManager trying to map enum type, which is not supported by the platform officially. I do not have this issue when using migrations / working with ORM only, so I assume it has to do with the SchemaManager itself. But please, a review is needed.

In the meantime I have made the SINGLE -> SCALAR change. I have also changed how the TypesDumper creates names - if no custom name is specified then nothing changes. But if custom name is used in config, then its pascal-cased and used instead of the base name of the enum

The tests I included in this PR are marked as skipped for the above mentioned issues

michnovka committed:
Fix unnecessary import

michnovka committed:
Fix FlagBag type always SCALAR

michnovka commented:
@ogizanagi Can we merge this?

michnovka committed:
Fix lint

michnovka commented:
@ogizanagi fixed lint errors, please re-run tests

ogizanagi commented on src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php:
```diff
@@ -37,6 +37,10 @@ private function addDoctrineDbalSection(ArrayNodeDefinition $rootNode)
             ->children()
+                ->booleanNode('enum_sql_declaration')
+                    ->defaultValue(false)
+                    ->info('If true, by default for enumerations, generate DBAL types with an ENUM SQL declaration with enum values instead of a VARCHAR/INT (Your platform must support it)')
```
```suggestion
                    ->info('If true, generate DBAL types with an ENUM SQL declaration with enum values instead of a VARCHAR/INT (Your platform must support it)')
```

ogizanagi commented on src/Bridge/Doctrine/DBAL/Types/AbstractEnumSQLDeclarationType.php:
```diff
@@ -0,0 +1,42 @@
+
+        $values = array_map(
+            function ($val) {
+                return "'" . $val->value . "'";
```
```suggestion
                return "'{$val->value}'";
```

ogizanagi commented on src/Bridge/Doctrine/DBAL/Types/AbstractEnumSQLDeclarationType.php:
```diff
@@ -0,0 +1,42 @@
+    public function getSQLDeclaration(array $fieldDeclaration, AbstractPlatform $platform): string
+    {
+        if (!$platform instanceof AbstractMySQLPlatform) {
+            throw new LogicException('ENUMs not supported on current platform');
```
```suggestion
            throw new LogicException('SQL ENUM type is not supported on the current platform');
```

ogizanagi commented on src/Bridge/Doctrine/DBAL/Types/TypesDumper.php:
```diff
@@ -49,8 +52,18 @@ public function getName(): string
+                break;
+            case self::TYPE_ENUM:
+                $baseClass = AbstractEnumSQLDeclarationType::class;
+                $this->appendDefaultOnNullMethods($code, $enumClass, $defaultOnNull);
```
Could be moved out of the switch

ogizanagi commented on src/Bridge/Doctrine/Common/AbstractTypesDumper.php:
```diff
@@ -35,7 +70,7 @@ private function dump(array $types): string
 
         $namespaces = [];
         foreach ($types as [$enumClass, $type, $name, $default]) {
-            $fqcn = self::getTypeClassname($enumClass, $type);
```
getTypeClassname unused now?

ogizanagi commented on src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php:
```diff
@@ -71,8 +75,16 @@ private function addDoctrineDbalSection(ArrayNodeDefinition $rootNode)
+                            ->info(<<<TXT
+Which column definition to use and the way the enumeration values are stored in the database:
+- scalar: VARCHAR/INT based on BackedEnum type
+- enum: ENUM(...values) - values are strings based on BackedEnum type (Your platform must support it)
```
```suggestion
- enum: ENUM(...values) as strings based on BackedEnum type (Your platform must support it)
```

ogizanagi commented on src/Bridge/Symfony/Bundle/DependencyInjection/ElaoEnumExtension.php:
```diff
@@ -113,9 +114,28 @@ private function prependDoctrineOdmConfig(array $config, ContainerBuilder $conta
+        $defaultStringType = $useEnumSQLDeclaration ? DBALTypesDumper::TYPE_ENUM : DBALTypesDumper::TYPE_SCALAR;
+
+        if (null === $type) {
+            $type = is_a($class, FlagBag::class, true) ? DBALTypesDumper::TYPE_SCALAR : $defaultStringType;
```
FlagBag should not be handled here, it'll have its dedicated type.
Let's change this method to use a `getDefaultType`

michnovka replied:
I do not understand what you mean here

ogizanagi commented on tests/Fixtures/Integration/Symfony/config/config.yml:
```diff
@@ -31,6 +31,9 @@ elao_enum:
         types:
             App\Enum\Suit:
                 default: !php/const App\Enum\Suit::Spades
+            #suit_sql_enum:
```
Let's remove until we add a MySQL integration test case (CI only)

ogizanagi commented on tests/Fixtures/Integration/Symfony/src/Entity/CardSQLEnum.php:
```diff
@@ -0,0 +1,45 @@
+
+// #[ORM\Entity]
+// #[ORM\Table(name: 'cards_sql_enum')]
+class CardSQLEnum
```
Same here

ogizanagi commented on tests/Integration/Bridge/Doctrine/DBAL/Type/EnumTypeTest.php:
```diff
@@ -53,6 +53,20 @@ public function testEnumType(): void
         self::assertSame(Suit::Hearts, $card->getSuit());
     }
 
+    public function testEnumSQLType(): void
```
Will be moved into a dedicated MySQL integration test case

ogizanagi commented on tests/Unit/Bridge/Doctrine/Common/Types/BaseTypesDumperTest.php:
```diff
@@ -55,4 +55,18 @@ private static function updateSnapshot(string $snapshotPath, string $content): v
+    public function testPascalCaseConvertor()
+    {
+        $testCases = [
+            'Foo\\Bar\\Baz' => 'FooBarBaz',
```
We should keep original namespace (and use a `@dataProvider`)

michnovka replied:
This test is just testing the "pascalization" of various strings. The actual algorithm that decides the name acts like you say (keeps namespace) but this is tested elsewhere by comparing the dumped config with some stub saved in tests folder

ogizanagi commented on tests/Unit/Bridge/Doctrine/Common/Types/BaseTypesDumperTest.php:
```diff
@@ -55,4 +55,18 @@ private static function updateSnapshot(string $snapshotPath, string $content): v
+    {
+        $testCases = [
+            'Foo\\Bar\\Baz' => 'FooBarBaz',
+            'Foo::Bar.baz' => 'FooBarBaz',
```
What use-cases are these test cases covering?

michnovka replied:
I want to test that All special chars are omitted and any small case letter following a nonalphanumeric char is converted to upper case

ogizanagi commented on tests/Unit/Bridge/Symfony/Bundle/DependencyInjection/ConfigurationTest.php:
```diff
@@ -161,6 +163,7 @@ private function getDefaultConfig(): array
     {
         return [
             'doctrine' => [
+                'enum_sql_declaration' => false,
```
Should add a test with true

ogizanagi commented on tests/Unit/Bridge/Doctrine/DBAL/Types/TypesDumperTest.php:
```diff
@@ -26,11 +26,11 @@ protected function getSnapshotPath(): string
+            ['Foo\Bar\BazWithDefault', 'scalar', 'baz_with_default', 'foo'],
+            ['Foo\Bar\Qux', 'scalar', 'Foo\Bar\Qux'],
+            ['Foo\Baz\Foo', 'scalar', 'foo'],
+            ['Foo\Baz\FooWithDefault', 'scalar', 'foo_with_default', 3],
```
A test case with an enum SQL declaration is missing

ogizanagi commented on tests/Unit/Bridge/Symfony/Bundle/DependencyInjection/ElaoEnumExtensionTest.php:
```diff
@@ -37,9 +37,9 @@ public function testDoctrineTypes(): void
-            [RequestStatus::class, 'single', 'request_status', 200],
+            [Suit::class, 'scalar', 'suit', null],
+            [Permissions::class, 'scalar', 'permissions', null],
+            [RequestStatus::class, 'scalar', 'request_status', 200],
```
A test case with `enum` is missing

michnovka committed:
Multiple changes based on @ogizanagi suggestions

michnovka committed:
Add TypesDumperTest enum case

michnovka committed:
Update XML schema for symfony type configuration. Updated tests

michnovka committed:
Fix doctrine_types_sql_enum configuration option and add tests covering XML and YAML declaration

michnovka committed:
MySQL test suite

michnovka committed:
Fix Doctrine deprecations fetch() and executeUpdate in tests

michnovka commented:
@ogizanagi I went over your comments and resolved most things. I also added some more tests which led me to find some bugs. All fixed.

I have created a MySQL test class. It can be enabled by changing `phpunit.xml.dist` to use MySQL instead of SQLite.

I still dont get your feedback on `FlagBag` inside `ElaoEnumExtension::resolveDbalType()`. The reason that check is there is to prevent using ENUM at any cost if FlagBag is used.

Please review once again and let me know if the way I enable MySQL tests is ok, or you wish to use some different approach. Right now I am checking `$_ENV['DOCTRINE_DBAL_URL']` for `pdo-mysql://`

michnovka commented:
I have a failing test on Windows due to DBAL URL I moved into `phpunit.xml.dist`

```xml
    <env name="DOCTRINE_DBAL_URL" value="sqlite:////%kernel.cache_dir%/db.sqlite" />
```

Anybody has windows who can test and tell me what it should be to work there? Thanks!

michnovka committed:
Fix DBAL URL for Windows

michnovka commented:
@ogizanagi I tried to fix it, can you re-run tests please?

michnovka committed:
Fix DBAL URL path

michnovka commented:
That didnt work well. lets see if this version works, seems I had an extra `/` in the first revision.

ogizanagi commented:
All green now ✅ 
Thank you @michnovka for your work and investigating the issues. I'll probably make some more minor tweaks to this PR in the next days, so I'm keeping it open for now as a reminder.

michnovka commented:
I think it would be a good idea to set up MySQL (or MariaDB) workflow test at least for Symfony 5.4 with PHP 8.1 on Linux

ogizanagi commented:
That's part of the plan indeed :)

michnovka committed:
Fix resolveDbalType to not check for FlagBag class

michnovka commented:
I finally understand the `FlagBag` issue. I can work to make `FlagBag` type as part of this PR, what do you think?

I also thought it might be nice to have an actual Symfony Form type of `FlagBag`, which would be a `select` with `multi=true`

ogizanagi commented:
Both of these features are listed in https://github.com/Elao/PhpEnums/issues/124 indeed.
If you'd like to work on it, that's great, but please open different PRs :)

michnovka commented:
Well given that I need lot of code from this PR and you plan to modify it
yet, I think Ill wait for the final merge

ogizanagi commented:
@michnovka : What do you mean? https://github.com/Elao/PhpEnums/pull/182 was merged

michnovka commented:
strange, I sent that email on April 15th :)
![Screenshot 2022-10-11 at 09-58-17 Re Elao_PhpEnums Add SQL Enum support (PR #179) - michnovka@gmail com - Gmail](https://user-images.githubusercontent.com/16553087/195032494-a92f4240-d81d-49c4-b2cc-01c0882dc456.png)
